AMP-AD Submodules
=================

Project Description
-------------------

Co-expression modules generated by Ben for transcriptomic data from the AMP-AD cohorts (10.7303/syn11932957.1). In this project, we generate submodules by pruning genes and dividing the broad modules into pathway-specific gene sets. We also generated quantitative traits for single-variant association mapping.

More information about where to find specific analyses or results can be found in `Summer2019Report.html` and `Summer2018Report.html`.

Environment
-----------

Scripts are written in either `R` or `Python`. I used `R` version 3.5.3 and `Python` version 3.7.3. Other important softwares used include:

1. `Inkscape` (0.92.2)
2. `Cytoscape` (3.7.1)
3. `RStudio` (1.2.1335)
4. `PLINK` (1.07)
5. `EMMAX` (2012-02-05 from http://csg.sph.umich.edu//kang/emmax/download/index.html)

Data Availability
-----------------

The full repository is not under version control. The complete project, including large and controlled files, is stored on Helix at `/sdata/carter-lab/carter/AMPAD/AMPAD_Submodules/`. 

Recommended Update Process
--------------------------

1. Clone repository onto local machine
2. Make changes
3. If changes involve large files, updated `.gitignore`
4. Push changes to version control
5. Run `./utils/update_sdata.sh` to update Helix with new contents
