AMP-AD Submodules
=================

Project Description
-------------------

This repository holds the scripts, RMarkdown files, and their html rendering from Nikhil Milind's summer project 2018-2019 in the Carter lab. The full project folder is in '/sdata/carter-lab/carter/AMPAD/AMPAD_Submodules/` on tier1. Results from this project are available on [Synapse](https://www.synapse.org/#!Synapse:syn23573590/wiki/607344).

Co-expression modules were generated by [Logsdon et al. (2019)](https://www.biorxiv.org/content/10.1101/510420v1) for transcriptomic data from the AMP-AD cohorts (10.7303/syn11932957.1). In this project, we generate submodules by pruning genes and dividing the broad modules into pathway-specific gene sets. We also generated quantitative traits for single-variant association mapping.

More information about where to find specific analyses or results can be found in `Summer2019Report.html` and `Summer2018Report.html`.

This project is now published in:   
Milind N, Preuss C, Haber A, Ananda G, Mukherjee S, John C, et al. (2020) Transcriptomic stratification of late-onset Alzheimer's cases reveals novel genetic modifiers of disease pathology. PLoS Genet 16(6): e1008775. https://doi.org/10.1371/journal.pgen.1008775

Environment
-----------

Scripts are written in either `R`(3.5.3) or `Python`(3.7.3). Other important softwares used include:

1. `Inkscape` (0.92.2)
2. `Cytoscape` (3.7.1)
3. `RStudio` (1.2.1335)
4. `PLINK` (1.07)
5. `EMMAX` (2012-02-05 from http://csg.sph.umich.edu//kang/emmax/download/index.html)

